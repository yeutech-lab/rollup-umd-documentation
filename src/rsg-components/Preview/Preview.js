import React, { Component } from 'react';
import PropTypes from 'prop-types';
import ReactDOM from 'react-dom';
import { transform } from 'buble';
import PlaygroundError from 'rsg-components/PlaygroundError';
import Wrapper from 'rsg-components/Wrapper';
import splitExampleCode from '../../utils/splitExampleCode';

/* eslint-disable no-invalid-this, react/no-multi-comp */

const Fragment = React.Fragment ? React.Fragment : 'div';
const FragmentTag = React.Fragment ? 'React.Fragment' : 'div';

const compileCode = (code, config) => transform(code, config).code;
const wrapCodeInFragment = code => `<${FragmentTag}>${code}</${FragmentTag}>;`;

// Wrap everything in a React component to leverage the state management
// of this component
class PreviewComponent extends Component {
	static propTypes = {
		component: PropTypes.func.isRequired,
		initialState: PropTypes.object.isRequired,
	};

	state = this.props.initialState;
	setStateBinded = this.setState.bind(this);

	render() {
		return this.props.component(this.state, this.setStateBinded);
	}
}

export default class Preview extends Component {
	static propTypes = {
		code: PropTypes.string.isRequired,
		evalInContext: PropTypes.func.isRequired,
	};
	static contextTypes = {
		config: PropTypes.object.isRequired,
		codeRevision: PropTypes.number.isRequired,
	};
	state = {
		error: null,
	};

	componentDidMount() {
		// Clear console after hot reload, do not clear on the first load
		// to keep any warnings
		if (this.context.codeRevision > 0) {
			// eslint-disable-next-line no-console
			console.clear();
		}

		this.executeCode();
	}

	shouldComponentUpdate(nextProps, nextState) {
		return this.state.error !== nextState.error || this.props.code !== nextProps.code;
	}

	componentDidUpdate(prevProps) {
		if (this.props.code !== prevProps.code) {
			this.executeCode();
		}
	}

	componentWillUnmount() {
		this.unmountPreview();
	}

	// Eval the code to extract the value of the initial state
	getExampleInitialState(compiledCode) {
		if (compiledCode.indexOf('initialState') === -1) {
			return {};
		}

		return this.props.evalInContext(`
			var state = {}, initialState = {};
			try {
				${compiledCode};
			} catch (err) {}
			return initialState;
		`)();
	}

	// Run example code and return the last top-level expression
	getExampleComponent(compiledCode) {
		return this.props.evalInContext(`
			var initialState = {};
			${compiledCode}
		`);
	}

	unmountPreview() {
		if (this.mountNode) {
			ReactDOM.unmountComponentAtNode(this.mountNode);
		}
	}

	executeCode() {
		this.setState({
			error: null,
		});

		const { code } = this.props;
		if (!code) {
			return;
		}

		const compiledCode = this.compileCode(code);
		if (!compiledCode) {
			return;
		}

		const { head, example } = splitExampleCode(compiledCode);
		const initialState = this.getExampleInitialState(head);
		const exampleComponent = this.getExampleComponent(example);
		const wrappedComponent = (
			<Wrapper onError={this.handleError}>
				<PreviewComponent component={exampleComponent} initialState={initialState} />
			</Wrapper>
		);

		window.requestAnimationFrame(() => {
			this.unmountPreview();
			try {
				ReactDOM.render(wrappedComponent, this.mountNode);
			} catch (err) {
				this.handleError(err);
			}
		});
	}

	compileCode(code) {
		try {
			const wrappedCode = code.trim().match(/^</) ? wrapCodeInFragment(code) : code;
			return compileCode(wrappedCode, this.context.config.compilerConfig);
		} catch (err) {
			this.handleError(err);
		}
		return false;
	}

	handleError = err => {
		this.unmountPreview();

		this.setState({
			error: err.toString(),
		});

		console.error(err); // eslint-disable-line no-console
	};

	render() {
		const { error } = this.state;
		return (
			<Fragment>
				<div ref={ref => (this.mountNode = ref)} />
				{error && <PlaygroundError message={error} />}
			</Fragment>
		);
	}
}
