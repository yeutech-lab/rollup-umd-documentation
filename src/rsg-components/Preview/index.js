export { default } from '../Preview/Preview';
