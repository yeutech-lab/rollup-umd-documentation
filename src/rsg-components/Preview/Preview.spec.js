import React from 'react';
import Preview from '../Preview';

/* eslint-disable no-console */

const evalInContext = a =>
	// eslint-disable-next-line no-new-func
	new Function('require', 'state', 'setState', 'const React = require("react");' + a).bind(
		null,
		require
	);
const code = '<button>OK</button>';
const options = {
	context: {
		config: {
			compilerConfig: {},
		},
		codeRevision: 0,
	},
};

const console$error = console.error;
const console$clear = console.clear;

afterEach(() => {
	console.error = console$error;
	console.clear = console$clear;
});

it('should unmount Wrapper component', () => {
	const actual = mount(<Preview code={code} evalInContext={evalInContext} />, options);

	expect(actual.html()).toMatch('<button');
	actual.unmount();
	expect(actual.html()).toBe(null);
});

it('should not not fail when Wrapper wasnâ€™t mounted', () => {
	console.error = jest.fn();

	const actual = mount(<Preview code="pizza" evalInContext={evalInContext} />, options);
	const node = actual.instance().mountNode;

	expect(console.error).toHaveBeenCalled();
	expect(node.innerHTML).toBe('');
	actual.unmount();
	expect(node.innerHTML).toBe('');
});

it('should wrap code in Fragment when it starts with <', () => {
	console.error = jest.fn();

	const actual = mount(<Preview code="<span /><span />" evalInContext={evalInContext} />, options);

	// If two spans weren't wrapped in a Fragment, we'd see an error in console
	expect(console.error).not.toHaveBeenCalled();
	expect(actual.html()).toMatchSnapshot();
});

it('should render component renderer', () => {
	const actual = shallow(<Preview code={code} evalInContext={evalInContext} />, {
		...options,
		disableLifecycleMethods: true,
	});

	expect(actual).toMatchSnapshot();
});

it('should not clear console on initial mount', () => {
	console.clear = jest.fn();
	mount(<Preview code={code} evalInContext={evalInContext} />, options);
	expect(console.clear).toHaveBeenCalledTimes(0);
});

it('should clear console on second mount', () => {
	console.clear = jest.fn();
	mount(<Preview code={code} evalInContext={evalInContext} />, {
		context: { ...options.context, codeRevision: 1 },
	});
	expect(console.clear).toHaveBeenCalledTimes(1);
});

it('should set initialState before the first render', () => {
	const code = `
initialState = {count:1};
<span>{state.count}</span>
	`;
	const actual = mount(<Preview code={code} evalInContext={evalInContext} />, options);
	expect(actual.html()).toMatchSnapshot();
});

it('should update state on setState', done => {
	const code = `
initialState = {count:1};
setTimeout(() => state.count === 1 && setState({count:2}));
<button>{state.count}</button>
	`;
	const actual = mount(<Preview code={code} evalInContext={evalInContext} />, options);

	actual
		.instance()
		.mountNode.querySelector('button')
		.click();

	setTimeout(() => {
		try {
			expect(actual.html()).toMatchSnapshot();
			done();
		} catch (err) {
			done.fail(err);
		}
	});
});
