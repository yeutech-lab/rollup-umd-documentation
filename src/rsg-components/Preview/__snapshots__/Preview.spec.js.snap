// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render component renderer 1`] = `
<React.Fragment>
  <div />
</React.Fragment>
`;

exports[`should set initialState before the first render 1`] = `

<div>
  <span>
    1
  </span>
</div>

`;

exports[`should update state on setState 1`] = `

<div>
  <button>
    2
  </button>
</div>

`;

exports[`should wrap code in Fragment when it starts with < 1`] = `

<div>
  <span>
  </span>
  <span>
  </span>
</div>

`;
